File: PlacesApp/PlacesShared/Place.swift
import Foundation

public struct Place: Codable, Identifiable, Hashable {
    public let id: String
    public let name: String
    public let address: String
    public var rating: Double?
    public var userRatingCount: Int?
    public var businessStatus: String?
    public var openNow: Bool?
    public var phone: String?
    public var weekdayHours: [String]?
    public var googleMapsUri: String?
    public var lastUpdated: Date?

    public init(
        id: String,
        name: String,
        address: String,
        rating: Double? = nil,
        userRatingCount: Int? = nil,
        businessStatus: String? = nil,
        openNow: Bool? = nil,
        phone: String? = nil,
        weekdayHours: [String]? = nil,
        googleMapsUri: String? = nil,
        lastUpdated: Date? = nil
    ) {
        self.id = id
        self.name = name
        self.address = address
        self.rating = rating
        self.userRatingCount = userRatingCount
        self.businessStatus = businessStatus
        self.openNow = openNow
        self.phone = phone
        self.weekdayHours = weekdayHours
        self.googleMapsUri = googleMapsUri
        self.lastUpdated = lastUpdated
    }
}

public struct SearchResult: Codable, Identifiable {
    public var id: String { placeId }
    public let placeId: String
    public let name: String
    public let address: String
    public let types: [String]
}

public struct PlaceDetails: Codable {
    public let placeId: String
    public let name: String
    public let address: String
    public let phone: String?
    public let rating: Double?
    public let userRatingCount: Int?
    public let businessStatus: String?
    public let openNow: Bool?
    public let weekdayHours: [String]?
    public let googleMapsUri: String?
}

File: worker/src/place.rs
use crate::models::{GooglePlace, PlaceDetails};

const FIELD_MASK: &str = "id,displayName,formattedAddress,nationalPhoneNumber,rating,userRatingCount,businessStatus,currentOpeningHours,regularOpeningHours,googleMapsUri";

pub async fn get_place_details(place_id: &str, api_key: &str) -> Result<PlaceDetails, String> {
    let url = format!("https://places.googleapis.com/v1/places/{place_id}");

    let mut init = worker::RequestInit::new();
    init.with_method(worker::Method::Get);

    let mut headers = worker::Headers::new();
    let _ = headers.set("X-Goog-Api-Key", api_key);
    let _ = headers.set("X-Goog-FieldMask", FIELD_MASK);
    init.with_headers(headers);

    let request = worker::Request::new_with_init(&url, &init)
        .map_err(|e| format!("Failed to create request: {e}"))?;

    let mut response = worker::Fetch::Request(request)
        .send()
        .await
        .map_err(|e| format!("Fetch failed: {e}"))?;

    if response.status_code() != 200 {
        let text = response.text().await.unwrap_or_default();
        return Err(format!(
            "Google API returned {}: {}",
            response.status_code(),
            text
        ));
    }

    let p: GooglePlace = response
        .json()
        .await
        .map_err(|e| format!("Failed to parse response: {e}"))?;

    let hours = p
        .current_opening_hours
        .as_ref()
        .or(p.regular_opening_hours.as_ref());

    Ok(PlaceDetails {
        place_id: p.id.unwrap_or_default(),
        name: p.display_name.map(|d| d.text).unwrap_or_default(),
        address: p.formatted_address.unwrap_or_default(),
        phone: p.national_phone_number,
        rating: p.rating,
        user_rating_count: p.user_rating_count,
        business_status: p.business_status,
        open_now: hours.and_then(|h| h.open_now),
        weekday_hours: hours.and_then(|h| h.weekday_descriptions.clone()),
        google_maps_uri: p.google_maps_uri,
    })
}
